### PR TITLE
MTL-1288 Fix Race Condition and Debug Mode

### PR DESCRIPTION
**Pull request by rustydb**
#### Summary and Scope

- Fixes MTL-1288
- Relates to MTL-1654 (removes contention with dmsquash-live)

##### Issue Type

- Bugfix Pull Request
- RFE Pull Request

###### Race Conditions Fixes

This adds support for preventing race conditions:

1. Add new paved function to report whether the pave has completed or not. This new function will return 0 or 1 if the wipe is in-progress or has completed, thus allowing dracut modules to wait on the pave before starting their functions.
2. Set `wait_for_dev` locks to wait for the OS disks to ensure dracut waits for this module to finish, preventing dmsquash-live from racing ahead when SQFSRAID is detected. Also remove the previously added labeling code that was working around the race condition, it only worked on the first deployment. Subsequent reboots were susceptible to the race condition. This fix mends the race condition on any boot.

###### Debug Mode

This also fixes missing or broken statements for enabling debug mode. Metal.debug was not honoring `metal.debug=1` in some cases. Ensure all files (except module-setup) honor `metal.debug`.

###### Consider NVME-Only Systems

Also create & use a new disks_exist function, this will ensure we wait for systems with `/dev/sd` or `/dev/nvme` block devices (a mix or pure). Otherwise it was possible to wait forever on systems with only NVME, meaning dracut only worked with mixed systems (SD + NVME, or SD only).

###### Safer Disk Boots

Make disk boots safer by moving the empty metal.server case to its own case that invokes `metal-md-scan`. This case is for disk boots, and only in this case will the mdscan function be used. This move helps separate context when `metal.server` is empty, setting it to empty will ensure none of the disk creation calls are made. When `metal.server` is empty we can assume that disks were already created. This is safer, since none of the disk creation code will load in a disk boot.

###### Misc.

- Add/update licenses.
- Fix broken warn statements, ensure quotes wrap metal_die and warn statements to prevent incorrect warnings.
- Add smarter fstab handling; merge old and new if different on reboot, and after validating they exist.
- When the partition table is unable to inform the kernel of changes, a system reset is performed

#### Prerequisites

- [x] I have included documentation in my PR (or it is not required)
- [x] I tested this on internal system (x) (if yes, please include results or a description of the test)
- [x] I tested rebooting with `metal.no-wipe=0`
- [x] I tested rebooting with `metal.no-wipe=1`
- [x] I tested rebooting with the disk boot loader
- [x] I tested the three scenarios above on k8s-masters, k8s-workers, and storage-ceph

#### Risks and Mitigations
 
What is less risky, or more risky now - or if your mod fails is there a new risk?

This mitigates the risk of loosing partitions from forgetting to manually wipe, ensuring that any boot from the PIT node will wipe the nodes.

This does pose some risk in-and-of itself being that this is significantly changes the codebase and that this codebase is hard to test.

### CODE REVIEW

**Review by jacobsalmela — APPROVED** [review]

**Comment by rustydb** [comment]
Tested on worker nodes. Waiting for an opportunity to test on masters.

I ran into one instance where `mdadm` was unable to use a partition because it was busy. I am seeing if I need to fix that or not, or if it was a fluke.

**Comment by rustydb** [comment]
This is working, the other modules are also working but conflicting one another.

The worker module is causing masters to fail booting because it's making masters wait on partitions that can't be created. It's an issue isolated to the worker module, nothing to do with this module.

**Comment by rustydb** [comment]
I'll fix the license checks before merging.

**Review by jacobsalmela — DISMISSED** [review]

**Comment by rustydb** [comment]
Working through testing.

File: 90metalmdsquash/metal-md-lib.sh [hunk 1]
```diff
@@ -294,16 +314,10 @@ add_sqfs() {
...
     mkdir -pv $sqfs_store
-    # if metal-nowipe=1, this will already exist
-    if mount -n -t xfs -L "${sqfs_drive_authority}" $sqfs_store; then
-        # nothing to do
-        umount $sqfs_store
-        return
-    elif mount -n -t xfs /dev/md/SQFS $sqfs_store; then
+    if mount -n -t xfs /dev/md/SQFS $sqfs_store; then
         mkdir -pv "$sqfs_store/$live_dir"
         fetch_sqfs "$sqfs_store/$live_dir" || metal_die 'Failed to fetch squashFS into squashFS storage!'
```
Comment (rustydb) [inline 90metalmdsquash/metal-md-lib.sh:319]: Debating if this should be `metal_die -b` to instantly reboot, or if it should remain as-is to facilitate investigations.

Comment (rustydb) [reply 90metalmdsquash/metal-md-lib.sh:319]: Probably not.. because I do want folks to be able to look and confirm that TCP/IP is broken.

**Comment by rustydb** [comment]
Done testing. NCNs wipe themselves whenever they're booted with `metal.no-wipe=0`, otherwise partitions are mounted normally when diskbooted or netbooted with `metal.no-wipe=1`.

**Pull request merged**